### PR TITLE
Tensor.scaled_dot_product_attention to match torch, used in LLaMA, and tested

### PR DESCRIPTION
**Pull request by jla524**
changes made (so far)
- implemented function
  - support is_causal and attn_mask
- wrote test cases
- use function in LLaMA

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/tensor.py                         3               0
------------------------------------------------------------
lines added in the tinygrad folder: 3
```

**Comment by geohot** [comment]
Nice to see you back! Bounty locked.

**Comment by jla524** [comment]
Thanks George! This is ready for review.

Here are the results for llama (how does Stacy have a 7900XTX?):

```
jacky@ephphatha tinygrad % python3 examples/llama.py 
using METAL backend
using LLaMA-7B model
...
User: Which is better, PyTorch or tinygrad?
Stacy: PyTorch is better than tinygrad because it's faster and bigger. [EOS]

User: I'm pretty sure tinygrad is faster. Have you tried running test_speed_v_torch.py?                 
Stacy: I'm sorry, but I'm not sure. I haven't tried it. [EOS]

User: You should try it, because tinygrad is faster on M1 and 7900XTX.
Stacy: I am sorry. I don't have an M1, but I have a 7900XTX. [EOS]
```

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -680,6 +680,12 @@ def dropout(self, p=0.5) -> Tensor:
     mask = (Tensor.rand(*self.shape, requires_grad=False) >= p).cast(dtypes.bool)
     return self * mask * (1/(1.0 - p))
 
+  @staticmethod
+  def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False) -> Tensor:
+    if is_causal: attn_mask = Tensor.ones(query.shape[-2], key.shape[-2]).tril(0).cast(dtypes.bool)
```
Comment (wozeparrot) [inline tinygrad/tensor.py:685]: Should probably `requires_grad=False` here

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -680,6 +680,12 @@ def dropout(self, p=0.5) -> Tensor:
     mask = (Tensor.rand(*self.shape, requires_grad=False) >= p).cast(dtypes.bool)
     return self * mask * (1/(1.0 - p))
 
+  @staticmethod
```
Comment (geohot) [inline tinygrad/tensor.py:683]: Instead of staticmethod, just make the query be `self`

Add type signatures to this function, and I think it's good!

**Comment by geohot** [comment]
Bounty paid!

**Pull request merged**